src/order/max.lean

```diff
@@ -66,33 +100,98 @@ several top elements. When `α` is linear, this is useful to make a case disjunc
+alias is_top_to_dual_iff ↔ _ is_bot.to_dual
+alias is_min_to_dual_iff ↔ _ is_max.to_dual
+alias is_max_to_dual_iff ↔ _ is_min.to_dual
+alias is_bot_of_dual_iff ↔ _ is_top.of_dual
+alias is_top_of_dual_iff ↔ _ is_bot.of_dual
+alias is_min_of_dual_iff ↔ _ is_max.of_dual
+alias is_max_of_dual_iff ↔ _ is_min.of_dual
 
 end has_le
 
 section preorder
 variables [preorder α] {a b : α}
 
-@[simp] lemma not_is_bot [no_min_order α] (a : α) : ¬is_bot a :=
-λ h, let ⟨b, hb⟩ := exists_lt a in hb.not_le $ h b
+lemma is_bot.mono (ha : is_bot a) (h : b ≤ a) : is_bot b := λ c, h.trans $ ha _
+lemma is_top.mono (ha : is_top a) (h : a ≤ b) : is_top b := λ c, (ha _).trans h
+lemma is_min.mono (ha : is_min a) (h : b ≤ a) : is_min b := λ c hc, h.trans $ ha $ hc.trans h
+lemma is_max.mono (ha : is_max a) (h : a ≤ b) : is_max b := λ c hc, (ha $ h.trans hc).trans h
 
-@[simp] lemma not_is_top [no_max_order α] (a : α) : ¬is_top a :=
-λ h, let ⟨b, hb⟩ := exists_gt a in hb.not_le $ h b
+lemma is_min.not_lt (h : is_min a) : ¬ b < a := λ hb, hb.not_le $ h hb.le
+lemma is_max.not_lt (h : is_max a) : ¬ a < b := λ hb, hb.not_le $ h hb.le
 
+lemma is_min_iff_forall_not_lt : is_min a ↔ ∀ b, ¬ b < a :=
+⟨λ h _, h.not_lt, λ h b hba, of_not_not $ λ hab, h _ $ hba.lt_of_not_le hab⟩
+
+lemma is_max_iff_forall_not_lt : is_max a ↔ ∀ b, ¬ a < b :=
+⟨λ h _, h.not_lt, λ h b hba, of_not_not $ λ hab, h _ $ hba.lt_of_not_le hab⟩
+
+@[simp] lemma not_is_min_iff : ¬ is_min a ↔ ∃ b, b < a :=
+by simp_rw [lt_iff_le_not_le, is_min, not_forall, exists_prop]
+
+@[simp] lemma not_is_max_iff : ¬ is_max a ↔ ∃ b, a < b :=
+by simp_rw [lt_iff_le_not_le, is_max, not_forall, exists_prop]
+
+@[simp] lemma not_is_min [no_min_order α] (a : α) : ¬ is_min a := not_is_min_iff.2 $ exists_lt a
+@[simp] lemma not_is_max [no_max_order α] (a : α) : ¬ is_max a := not_is_max_iff.2 $ exists_gt a
+
+namespace subsingleton
+variable [subsingleton α]
+
+protected lemma is_bot (a : α) : is_bot a := λ _, (subsingleton.elim _ _).le
+protected lemma is_top (a : α) : is_top a := λ _, (subsingleton.elim _ _).le
+protected lemma is_min (a : α) : is_min a := (subsingleton.is_bot _).is_min
+protected lemma is_max (a : α) : is_max a := (subsingleton.is_top _).is_max
+
+end subsingleton
 end preorder
 
 section partial_order
 variables [partial_order α] {a b : α}
 
+protected lemma is_min.eq_of_le (ha : is_min a) (h : b ≤ a) : b = a := h.antisymm $ ha h
+protected lemma is_min.eq_of_ge (ha : is_min a) (h : b ≤ a) : a = b := h.antisymm' $ ha h
+protected lemma is_max.eq_of_le (ha : is_max a) (h : a ≤ b) : a = b := h.antisymm $ ha h
+protected lemma is_max.eq_of_ge (ha : is_max a) (h : a ≤ b) : b = a := h.antisymm' $ ha h
+
+--TODO: Delete in favor of the above
 lemma is_bot.unique (ha : is_bot a) (hb : b ≤ a) : a = b := (ha b).antisymm hb
-lemma is_top.unique (ha : is_top a) (hb : a ≤ b) : a = b := le_antisymm hb (ha b)
+lemma is_top.unique (ha : is_top a) (hb : a ≤ b) : a = b := hb.antisymm (ha b)
```

Comment:
I don't mind if this waits for a future PR, but I would expect a lemma `ha.unique` to have as its other hypothesis something like `hb : is_top b`. And this one would have a name more like `is_max.eq_of_le` already does.

Reply:
I totally agree but fixing this scope creeps and is part of why you thought #11408 was too wide. I specifically reverted this change.

Reply:
Yeah, totally fine for this to be in a different PR!